### PR TITLE
feat(rules): Support keyed grouping of rule futures

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Supports specifying a key name for the rule callback. This allows for callbacks that are bound at runtime by specifying a key for them.

### CODE REVIEW

**Review by tkaemming — CHANGES_REQUESTED** [review]
This generally makes sense, but I think the documentation bit would be beneficial and would like confirmation that the introduction of the `key` parameter isn't going to collide with any existing known usage.

File: src/sentry/rules/base.py [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
 from sentry.utils.html import escape
 
-CallbackFuture = namedtuple('CallbackFuture', ['callback', 'kwargs'])
+CallbackFuture = namedtuple('CallbackFuture', ['callback', 'key', 'kwargs'])
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/rules/base.py]: ea4e2f1e4

File: src/sentry/rules/processor.py [hunk 2]
```diff
@@ -138,11 +139,16 @@ def apply_rule(self, rule):
                 continue
 
             for future in results:
-                self.futures_by_cb[future.callback
-                                   ].append(RuleFuture(rule=rule, kwargs=future.kwargs))
+                key = future.key or future.callback
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/rules/processor.py]: 7df849e50

File: src/sentry/rules/base.py [hunk 3]
```diff
@@ -105,9 +105,10 @@ def validate_form(self):
 
         return form.is_valid()
 
-    def future(self, callback, **kwargs):
+    def future(self, callback, key=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (EvanPurkhiser) [reply src/sentry/rules/base.py]: Yeah this isn't great. As far as I can tell none of our actions that generate futures actually pass _any_ other arguments besides the callback. 3rd party plugins *could* be doing this, but I'm not even aware of any plugins that add additional rule actions.

**Comment by EvanPurkhiser** [comment]
Updated for 2nd pass review @tkaemming. Thanks

**Review by tkaemming — APPROVED** [review]
Godspeed

**Pull request merged**